_posts/2015-01-30-cloudpipe.markdown

```diff
@@ -0,0 +1,103 @@
+
+Once you have an account running on a cluster somewhere, you can install multyvac with:
+
+```bash
+pip install multyvac
+```
+
+Then, you can submit functions as jobs:
+
+```python
+import multyvac
+
+# Connect to the cluster
+multyvac.config.set_key(api_key='username',
+                        api_secret_key='54321',
+                        api_url='https://somewhere/v1')
+
+def add(a, b):
+  return a + b
+
+# Submit the function as a job. Args and kwargs will be
+# passed to it server-side.
+jid = multyvac.submit(add, 3, 4)
+
+# Use the job ID to get a Job object...
+job = multyvac.get(jid)
+
+# ... and use the Job object to get the result.
+# This call will block until the result is ready.
+result = job.get_result()
+```
+
+Multyvac uses some neat pickling tricks to send your function over the wire: you can actually send closures, or code that references external libraries, and it'll work. Other languages' SDKs may not be able to pull the same range of tricks, depending on how evil each runtime lets us be, but they'll support the same basic idea.
+
+One important thing to notice is that the way work is submitted to the API is completely language-agnostic: it's a contract between the SDK and the base Docker image that's invoked on the server. The Cloudpipe API speaks strictly in terms of POSIX things: command and arguments, environment variables, and standard input, output and error streams.
+
+## How it works
+
+Behind that URL, Cloudpipe is built from a small set of parts, all written in Go.
+
+*[cloudpipe/cloudpipe](https://github.com/cloudpipe/cloudpipe)*
+
+The **API server** is an HTTPS server that hosts a JSON API. It's a fairly thin layer that performs  CRUD operations on job models persisted in a MongoDB instance.
+
+The **job runner** is a separate goroutine within the same process as each API server. It polls a queue of submitted jobs in MongoDB and uses the [go-dockerclient library](https://github.com/fsouza/go-dockerclient) to launch each one. The runner creates a container from the requested image, attaches to its stdin, stdout and stderr, and starts it. Any provided stdin is fed in to the process; stdout and stderr are collected and used to update the job's model in Mongo as it executes.
+
+In development mode, the container is launched on the same server that you're running everything else on, but in any "real" setting user jobs should run on a different machine. We're planning to use [swarm](https://github.com/docker/swarm) as a way to manage submitting jobs to a cluster.
+
+*[cloudpipe/auth-store](https://github.com/cloudpipe/auth-store)*
+
+Under the hood, the API server talks to a pluggable **authentication server** over HTTPS on an internal network, secured with a closed network of TLS credentials -- meaning, the authentication server only accepts connections from clients who offer certificates signed by an internal certificate authority that we (the cluster operators) control.
```

Comment:
I actually started a section on security and yanked it because it was too big of a topic. Definitely another post. And the key generation could be another, right. Guess I have some writing to do!